src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/EnumConverter.cs

```diff
@@ -74,11 +85,27 @@ public EnumConverter(EnumConverterOptions converterOptions, JsonNamingPolicy? na
                 ulong key = ConvertToUInt64(value);
                 string name = names[i];
 
-                _nameCache.TryAdd(
-                    key,
-                    namingPolicy == null
-                        ? JsonEncodedText.Encode(name, encoder)
-                        : FormatEnumValue(name, encoder));
+                if (namingPolicy == null)
+                {
+                    if (_namingPolicyCache.Count < NameCacheSizeSoftLimit)
+                    {
+                        _namingPolicyCache.TryAdd(key, JsonEncodedText.Encode(name, encoder));
+                    }
+                }
+                else
+                {
+                    string formatted = FormatEnumValueToString(name);
+
+                    if (_namingPolicyCache.Count < NameCacheSizeSoftLimit)
+                    {
+                        _namingPolicyCache.TryAdd(key, JsonEncodedText.Encode(formatted, encoder));
+                    }
+
+                    // We don't enforce a cap on the number on the deserialization cap since we need to
```

Comment:
Do we need to enforce a cap `_namingPolicyCache` either? As long as the number is bounded by the number of fields in the enum I think it's fair game. The primary purpose of the cap is to defend against a combinatorial explosion when needing to deal with arbitrary combinations of comma-separated values, which can be exploited by an attacker. 

So it feels to me that the code might as well look as follows:
```C#
string jsonName = namingPolicy is null ? name : FormatEnumValueToString(name);
_namingPolicyCache.TryAdd(key, JsonEncodedText.Encode(jsonName, encoder));
_enumValueCache?.TryAdd(jsonName, value);
```